optuna/pruners/percentile.py

```diff
@@ -93,12 +98,29 @@ def prune(self, study, trial):
         if step is None:
             return False
 
-        if step <= self.n_warmup_steps:
+        n_warmup_steps = self.n_warmup_steps
+        if step <= n_warmup_steps:
             return False
 
-        if len(trial.intermediate_values) == 0:
+        intermediate_values = trial.intermediate_values
+        if len(intermediate_values) == 0:
             return False
 
+        interval_steps = self.interval_steps
+        base_index = 1
+        lower_bound = (
+            ((step - n_warmup_steps - base_index) // interval_steps) * interval_steps
+            + n_warmup_steps + base_index)
+        is_above_bound = False
+        for s in sorted(intermediate_values.keys(), reverse=True):
```

Comment:
This code takes `O(N log N)` steps (where `N = len(intermediate_values)`).
I think that we can reduce the complexity to `O(N + M log M)` (where `M` is the number of keys that are greater than or equal to `lower_bound`.; typically `M` is a very small number) by filtering `intermediate_values` in advance such as `sorted([s for s in intermediate_values.key() if s >= lower_bound], reverse=True)`.

Reply:
Thanks you for pointing out, I noticed we can do this in `O(n)` since we just have to find the second last step, comparing it to the lower bound.

Reply:
💯